### PR TITLE
fix(reachability): test non-HTTP Services with TCP only

### PR DESCRIPTION
**Pull request by nadaverell**
## Summary

The Reachability tab's in-cluster test sent `tcp,http` to every route — including Postgres, Redis/Valkey, MongoDB and Kafka ports the static prober already knows are non-HTTP. That produced misleading results and sent noisy, unrelated HTTP bytes at non-HTTP servers. Non-HTTP Service ports now get **TCP-only** probes end to end: an honest transport check instead of a mis-probe, and no probe at all for protocols Radar can't speak (UDP/SCTP stay explicit gaps).

## What changed

**Every in-cluster request carries a protocol.** `ProbeRequest` gains `protocol` (`http` | `https` | `tcp`), classified from the declared Service-port signals (protocol, name, appProtocol — including the non-HTTP-on-443 edge case and Valkey-named ports alongside Redis). TCP requests omit the HTTP-only scheme/host/path fields, ignore HTTP path overrides, and the copyable fallback argv matches the Job exactly — no HTTP flags for a TCP-only run. UDP/SCTP ports classify to no request at all.

**Non-HTTP Service ports are testable, not invisible.** The laptop's API-server vantage skips non-HTTP ports by design, which previously left a direct non-HTTP Service with no route for the in-cluster runner to act on. A vantage-skipped Service port now materializes as an honest `not-tested` route carrying a TCP candidate; the in-cluster Job dials it, and a live pass folds to `reached` with real confidence. Materialization is scoped to Service subjects, so an untested backend can never erase a still-untested front door.

**Dual-protocol port numbers resolve to the TCP sibling.** When TCP and UDP share a number (kube-dns declares both `:53`), the request candidate belongs to the TCP-declared entry — requests only travel TCP, so this is a deliberate assignment, never declaration-order luck. With no TCP interpretation among the matches, the shape fails closed. A TCP pass never erases the UDP/SCTP sibling's unsupported-protocol gap, and stale vantage skips are pruned per-port only when every route on that port genuinely resolved.

**Skip semantics say what each gap means.** The inapplicable HTTP proxy beside a completed direct TCP dial is benign (nothing was lost); HTTPS remains a vantage gap until an application-layer request actually completes. A Service scaled to zero keeps its intentional-dormancy framing: the benign route itself carries "no running backends (scaled to 0)" and no probe is recommended against a backend that cannot answer.

**The consent gate names the traffic the Job actually sends.** A TCP-only candidate reads "TCP connection to database:6379" on the mutation-consent dialog — never a fabricated `GET http://…/` — and path overrides apply to HTTP(S) requests only. A retained UDP/SCTP sibling beside a TCP candidate counts as its own coverage gap, never absorbed by the TCP route's result.

**The UI never claims a request that didn't run.** `completedRequestMode` derives what actually completed (HTTP / TCP / mixed / none) from the folded trace, and `RequestIndicator` renders each honestly — "tested with TCP connection", "no application request completed from here" — instead of a blanket "tested with GET /". The HTTP path editor hides on TCP-only subjects, and the MCP `diagnose` schema and docs state the same contract for agents.

## Testing

- `go test ./...` (trace, reachability, server, mcp), `go vet`, `make tsc`, full vitest suites (k8s-ui trace + web) — green on the current base.
- Live validation on a disposable kind cluster with Valkey 8.1 and Postgres 17: each Service surfaced one honest `not-tested` TCP route; each in-cluster Job returned exactly one successful `tcp` result folding to `reached`/real (passed=1, skipped=0); Valkey logs stayed free of HTTP/TLS noise; the self-deleting Jobs left nothing behind.
- Visual-test capture skipped: the protocol labels and TCP-only path-control behavior are pinned by rendered-component tests; a browser smoke covered the surrounding reachability view.

**ExternalName aliases state their boundary.** Radar's in-cluster probes never dial external aliases, so the split-horizon and internal-address skips no longer prescribe the in-cluster test - they hand a copyable `kubectl run … nslookup` / `nc -vz` one-liner to check the alias from a pod by hand.

## Notes

Deliberately excluded (future work): native protocol probing (a real redis PING, gRPC health check), UDP/SCTP support, runtime-probe catalogs, target-container exec, ephemeral containers, and any new permissions. This PR is only the protocol-honesty layer those would build on.

---

> [!NOTE]
> **Medium Risk**
> Touches reachability verdicts, in-cluster probe Jobs, and coverage counting—operator-facing diagnosis—but changes are protocol-scoped with broad tests and no new RBAC.
> 
> **Overview**
> Reachability and diagnose no longer send **HTTP probes to ports that are clearly non-HTTP** (Redis/Valkey, Postgres, etc.). Each route’s in-cluster request now carries a **`protocol`** (`http` | `https` | `tcp`); TCP jobs dial with **`tcp` only**, strip HTTP fields, ignore path overrides, and copyable fallback commands match that shape.
> 
> **Service subjects** can still be exercised in-cluster when the laptop vantage only skipped the port: those skips become honest **`not-tested` routes** with TCP candidates that fold to **`reached`** on success. **Coverage/skip math** is tightened for multi-protocol ports (e.g. TCP+UDP on `:53`), transport-only reaches vs HTTPS app-layer gaps, scale-to-zero benign routes, and port-key parsing for named skip labels.
> 
> **UI and consent** describe what actually ran (`completedRequestMode`, TCP vs GET copy, hide path editor on TCP-only subjects) and name mutation consent as **“TCP connection to …”** for TCP routes. Docs/MCP schema and ExternalName skip hints align with the same contract.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 12bd96c3d7e01132d1bc01a85a84d7a5c41175ec. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: internal/trace/coverage.go [hunk 1]
```diff
@@ -1828,11 +1829,22 @@ func attachInClusterRequest(routes []RouteResult, host, path string, cfg *HopCon
 // guesses the leading literal and flags PathGuessed - the UI surfaces that and
 // lets the user correct it before running.
 func guessInClusterRequest(host, path string, port PortMap) ProbeRequest {
-	req := ProbeRequest{Scheme: schemeForPort(port), Host: concreteHost(host)}
+	req := ProbeRequest{
+		Protocol: protocolForPort(port),
+		Scheme:   schemeForPort(port),
+		Host:     concreteHost(host),
+	}
```
[the remark this answers is not in the document]
Comment (nadaverell) [reply internal/trace/coverage.go:2503]: Good catch; this was a real edge case. The runner now derives the executable scheme from `Protocol` and passes no scheme at all for TCP-only requests, so a non-HTTP Service on port 443 cannot trigger TLS. The regression test covers `protocol=tcp`, `scheme=https`, port 443 and asserts both `layers=tcp` and an empty runner scheme. Fixed in 348d8ff6.

File: internal/trace/coverage.go [hunk 2]
```diff
@@ -1514,6 +1543,12 @@ func routeFromProbes(routeID, target string, probes []probe.Result) (RouteResult
 		}
 	}
 	if len(real) == 0 && len(indirect) == 0 {
+		if vantageSkip != nil {
+			return RouteResult{
+				Route: routeID, Target: target, Outcome: OutcomeNotTested,
+				Evidence: vantageSkip.Reason, Command: vantageSkip.Command,
+			}, true
+		}
```
[the remark this answers is not in the document]
Comment (nadaverell) [reply internal/trace/coverage.go]: Fixed in 032897fc. Vantage-only route materialization is now scoped to Service subjects, so ExternalName/backend/front-door skip shapes remain unchanged. Service skips also retain their port identity, and regression tests pin both the Service-only materialization and the ExternalName/non-Service exclusion.

File: internal/trace/probes.go [hunk 3]
```diff
@@ -1127,6 +1141,7 @@ func probePodsByName(ctx context.Context, h *Hop, vantage probe.Vantage, client
 					"HTTPS backend - the API-server proxy speaks plain HTTP and can't verify TLS on this port. Test it directly.",
 					portForwardCmd("pod", h.Resource.Namespace, name, cp.Port)+fmt.Sprintf("   # then: curl -k https://localhost:%d/", cp.Port))
 				skip.Path = probe.PathAPIServer
+				skip.Port = cp.Port
```
[the remark this answers is not in the document]
Comment (nadaverell) [reply internal/trace/probes.go:1253]: Fixed in 4f9c6877. Pod-proxy skips now carry structured protocol-aware classes: non-HTTP is benign only when the same pod hop had a direct in-cluster TCP path, otherwise vantage; HTTPS remains vantage because TCP alone does not verify TLS/application behavior; UDP/SCTP remain coverage. Tests pin the production-shaped Service+ready-Pod result (passed=1, skipped=0), the in-cluster/no-PodIP fail-closed branch, HTTPS, and UDP/SCTP.

File: internal/trace/coverage.go [hunk 4]
```diff
@@ -1244,6 +1276,16 @@ func routesByPort(routeID, backendName, fallbackTarget string, probes []probe.Re
...
 		r, ok := routeFromProbes(rid, target, ps)
 		if !ok {
+			if materializeVantageSkips {
+				for _, p := range ps {
+					if p.Skipped && skipClassOf(p) == SkipClassVantage {
+						return RouteResult{
+							Route: rid, Target: target, Outcome: OutcomeNotTested,
+							Evidence: p.Reason, Command: p.Command,
+						}, true
+					}
```
[the remark this answers is not in the document]
Comment (nadaverell) [reply internal/trace/coverage.go:1732]: Fixed in 8b99d7d3. An authoritative scale-to-zero finding now normalizes a materialized not-tested route to the existing benign unreachable/dormant state, so the in-cluster runner excludes it. Matching Service-port skip context is marked benign, rewritten to explain that no backends are running, and has its impossible port-forward command removed. Port scoping preserves unrelated gaps and commands; tests pin both shapes.

File: internal/reachability/incluster.go [hunk 5]
```diff
@@ -162,8 +164,8 @@ func runInClusterTests(ctx context.Context, typed kubernetes.Interface, image st
 		}
 		opts := RunOptions{
 			Image: image, Namespace: namespace, Target: dialTarget,
-			Scheme: req.Scheme, Host: req.Host, Path: req.Path,
-			Layers: "tcp,http",
+			Scheme: schemeForProtocol(req.Protocol), Host: req.Host, Path: req.Path,
+			Layers: layersForProtocol(req.Protocol),
```
[the remark this answers is not in the document]
Comment (nadaverell) [reply internal/reachability/incluster.go:168]: Fixed in 1a4e2533. FallbackCommand now mirrors the Job argv: it omits --scheme when Scheme is empty, preserves layers=tcp, and includes the same 3s probe timeout. The TCP regression asserts no scheme/host/path are present.

File: packages/k8s-ui/src/components/trace/ReachabilityExplainer.tsx [hunk 6]
```diff
@@ -318,19 +336,45 @@ export function ReachabilityExplainer({ trace, probed, inClusterRunning, probePa
...
   const visibleRows = collapsible && !expanded ? rows.slice(0, COLLAPSED_ROWS) : rows
   const hiddenCount = rows.length - COLLAPSED_ROWS
+  const rowProbes = (row: MatrixRow) => Object.values(row.cells).filter((probe): probe is ProbeResult => !!probe)
+  const rowMethod = (row: MatrixRow) => {
+    if (row.protocol === 'tcp') return 'TCP'
+    if (row.protocol === 'http' || row.protocol === 'https') return 'GET'
+    if (rowProbes(row).some((probe) => !probe.skipped && probe.layer === 'http')) return 'GET'
+    if (rowProbes(row).some((probe) => !probe.skipped && (probe.layer === 'tcp' || probe.layer === 'tls'))) return 'TCP'
+    return 'CHECK'
+  }
```
[the remark this answers is not in the document]
Comment (nadaverell) [reply packages/k8s-ui/src/components/trace/ReachabilityExplainer.tsx]: Fixed in 1a4e2533. Row badges now prefer non-skipped evidence before intended protocol: completed HTTP is GET, otherwise completed TCP/TLS is TCP, with route protocol used only when nothing ran. The rendered HTTPS preliminary-TCP case now pins a TCP badge and no GET badge.
